lib/utils/auth.js

```diff
@@ -20,4 +25,49 @@ const generateBasicAuthToken = (clientId, clientSecret) => ({
   Authorization: `Basic ${Buffer.from(`${clientId}:${clientSecret}`).toString('base64')}`,
 });
 
-module.exports = { generateBasicAuthToken };
+const getV5Token = async () =>
+  axios.post(
+    `${config.get('piiServerUrl')}/api/v5/auth/token`,
+    {
+      grant_type: 'client_credentials',
+    },
+    {
+      auth: {
+        username: config.get('clientId'),
+        password: config.get('clientSecret'),
+      },
+    }
+  );
+// give a 10 second buffer for any network latency i.e. CA => AU
+const TOKEN_EXPIRATION_BUFFER = 10000;
+
+const setAxiosTokenData = (axiosInstance, { expires_in: expiresIn, created_at: createdAt }) => {
+  const expiresAt = add(createdAt, expiresIn * 1000 - TOKEN_EXPIRATION_BUFFER);
+
+  axiosInstance.defaults.tokenData = { expiresIn, createdAt, expiresAt };
+};
+
+const isTokenValid = dateObj => get(axios, 'defaults.tokenData.expiresAt') > dateObj.now();
+
+/**
+ * Set axios' default auth header to OAuth Bearer Token from v5 RO API
+ *
+ * @async
+ * @param {object} [dateObj=Date] Date object used to validate the token expiry date.
+ *
+ * @example
+ * await setV5Token()
+ */
+const setV5Token = async (dateObj = Date) => {
+  if (isTokenValid(dateObj)) {
+    return;
+  }
+
+  const { data } = await getV5Token();
+
+  setAxiosTokenData(axios, data);
+
+  axios.defaults.headers.common.Authorization = `Bearer ${data.access_token}`;
+};
+
+module.exports = { generateBasicAuthToken, setV5Token };
```

Comment:
Concerns with having two separate token setters (one here, the other [for V4] in `lib/auth.js`):

- it won't always be clear which one to use
- the V4 strategy uses an in-flight lock mitigation strategy, which isn't used here; what happens if several auth token requests are made and there's a race condition?
- we don't have a retry strategy here, either 👀 

I think we'd be wise to merge the calls together…

Reply:
yeah, I was considering using the `interceptor` axios API but given the response time (~3s) to get the token, I went with the above implementation.

Thoughts? https://thedutchlab.com/blog/using-axios-interceptors-for-refreshing-your-api-token